scheduler-pool/benches/scheduler.rs

```diff
@@ -0,0 +1,894 @@
+            if pending_transactions.is_empty() {
+                return Ok(());
+            }
+
+            let mut tx_account_locks: Vec<_> = Vec::with_capacity(pending_transactions.len());
+            for tx in pending_transactions {
+                tx_account_locks
+                    .push(tx.get_account_locks(bank.get_transaction_account_lock_limit())?);
+            }
+
+            // the dependency graph contains the indices that must be executed (marked with
+            // State::Done) before they can be executed
+            let dependency_graph = Self::build_dependency_graph(&tx_account_locks);
+
+            #[derive(Clone)]
+            enum State {
+                Blocked,
+                Processing,
+                Done,
+            }
+
+            let mut processing_states: Vec<State> = vec![State::Blocked; dependency_graph.len()];
+            let mut signature_indices: HashMap<&Signature, usize> =
+                HashMap::with_capacity(dependency_graph.len());
+            signature_indices.extend(
+                pending_transactions
+                    .iter()
+                    .enumerate()
+                    .map(|(idx, tx)| (tx.signature(), idx)),
+            );
+
+            let mut is_done = false;
+            while !is_done {
+                is_done = true;
+                for idx in 0..processing_states.len() {
+                    match processing_states[idx] {
+                        State::Blocked => {
+                            is_done = false;
+
+                            // if all the dependent txs are executed, this transaction can be
+                            // scheduled for execution.
+                            if dependency_graph[idx]
+                                .iter()
+                                .all(|idx| matches!(processing_states[*idx], State::Done))
+                            {
+                                self.inner_scheduelr.schedule_execution(Arc::new((
+                                    pending_transactions[idx].clone(),
+                                    idx,
+                                )));
+                                // this idx can be scheduled and moved to processing
+                                processing_states[idx] = State::Processing;
+                            }
+                        }
+                        State::Processing => {
+                            is_done = false;
+                        }
+                        State::Done => {}
+                    }
+                }
+
+                if is_done {
+                    break;
+                }
+
+                let mut executor_responses: Vec<_> = vec![receiver.recv().unwrap()];
+                executor_responses.extend(receiver.try_iter());
+                for r in &executor_responses {
+                    processing_states[*signature_indices.get(r).unwrap()] = State::Done;
+                }
+            }
+            Ok(())
```

Comment:
&lt;bike-shedding&gt;

I wonder if instead of going through all `dependency_graph.len()` entries of the vec every time, and doing a search for the signature via a hash map it would make sense to just have a few mutating vectors.

Something like

```rust
            #[derive(Copy, PartialEq, Eq)]
            enum State {
                Blocked,
                Processing,
                Done,
            }

            let mut state = vec![State::Blocked; dependency_graph.len()];
            let mut blocked = pending_transactions
                .iter()
                .enumerate()
                .map(|(idx, tx)| (idx, tx.signature()))
                .collect<Vec<_>>();

            let mut pending = Vec::with_capacity(dependency_graph.len());

            while !(blocked.is_empty() && pending.is_empty()) {
                for blocked_i in 0..blocked.len() {
                    if blocked_i >= blocked.len() {
                        break;
                    }

                    let (tx_idx, _signature) = blocked[blocked_i];
                    // if all the dependent txs are executed, this transaction can be
                    // scheduled for execution.
                    if dependency_graph[tx_idx]
                        .iter()
                        .all(|dependency_idx| state[*dependency_idx] == State::Done)
                    {
                        self.inner_scheduelr.schedule_execution(Arc::new((
                            pending_transactions[tx_idx].clone(),
                            tx_idx,
                        )));

                        pending.push(blocked.swap_remove(blocked_i));
                        state[tx_idx] = State::Processing;
                    }
                }

                assert!(
                    !pending.is_empty(),
                    "If there is no pending work but `blocked` is not empty we \
                     are in a deadlock"
                );

                let done_tx_sig = receiver.recv()
                    .expect("`pending` is non-empty");
                let pending_i = pending
                    .position(|(_idx, signature) signature == done_tx_sig)
                    .expect("`pending` must contain the completed transaction");

                let done_tx_idx = pending.swap_remove(pending_i).0;
                state[done_tx_idx] = State::Done;
            }
            Ok(())
        }
```

Considering it is testing code, not sure if it really matters.

Reply:
> I wonder if instead of going through all `dependency_graph.len()` entries of the vec every time, and doing a search for the signature via a hash map it would make sense to just have a few mutating vectors.
> 
> Something like ...

yeah, it looks competitive with current algo, at least. cc: @buffalu possible algo improve for fast replay.

the `.position` part is linear scan but should fare well with `HashMap` as `pending_transactions` shouldn't be that large to begin with.

also, i think reviving `receiver.try_iter()` might worth.

> <bike-shedding>
> ...
> Considering it is testing code, not sure if it really matters.

np. thanks for suggestion. I leave these code as-is as you pointed out that it is bench code (the bench isn't trying to eval this code; it just want small scheduler logic). 

but i just cc-ed people who care about it. :)